lib/puppet/provider/service/init.rb

```diff
@@ -21,7 +21,7 @@ def self.defpath
   end
 
   # Debian and Ubuntu should use the Debian provider.
-  confine :false => %w[Debian Ubuntu].include?(Puppet.runtime[:facter].value('os.name'))
+  confine :false => %w[Debian Ubuntu Cumulus].include?(Puppet.runtime[:facter].value('os.name'))
```

Comment:
We already set `systemd` as the default provider in https://github.com/puppetlabs/puppet/blob/dc0e3baae305a6d25164771661edb742ede656ee/lib/puppet/provider/service/systemd.rb#L33 so curious why we need to "exclude" `init`? Is the `os.name` incorrect?

Reply:
This is what is what fater returns on Cumulus:

```
root@edd-sw30:~# facter -p os.name
Cumulus
root@edd-sw30:~#
```

The exclusion solution was suggested somewhere on serverfault and is working for me.

Reply:
> The exclusion solution was suggested somewhere on serverfault 

The pattern we've been following is to specify `systemd` as the default for a given platform and then exclude older versions in situations where the platform uses `init` or something else. For example, we do this for `debian`

https://github.com/puppetlabs/puppet/blob/dc0e3baae305a6d25164771661edb742ede656ee/lib/puppet/provider/service/systemd.rb#L27-L28

This way it's future compatible (so long as `systemd` remains the default) and in theory it helps when documenting which providers are enabled by default on each platform. If we just exclude Cumulus from `init`, it's not clear what provider is "next" precedence-wise.

I think the root issue is this line is expecting the `os.name` fact to return `cumuluslinux` (case-insensitively) https://github.com/puppetlabs/puppet/blob/dc0e3baae305a6d25164771661edb742ede656ee/lib/puppet/provider/service/systemd.rb#L33

Maybe it did that in previous versions of Cumulus? Or maybe facter 4 is reporting the `os.name` fact differently on Cumulus? What is the output of `facter os`? Could you try installing puppet-agent 6 on Cumulus and see what `facter os` reports?

Related commits

15ed9d149d216697269e9ee8ce32630c12cc00e1
686d5ca1b3e3390bc5f5e4c3665427ebe3620801
7ba4df54c30f77702728ec5bbc98c8ffa2319c41

Reply:
root@edd-sw30:~# puppet --version
7.23.0

root@edd-sw30:~# facter os
{
  architecture => "amd64",
  distro => {
    codename => "cumulus",
    description => "Cumulus Linux",
    id => "Cumulus-linux",
    release => {
      full => "12.5",
      major => "12",
      minor => "5"
    }
  },
  family => "Debian",
  hardware => "x86_64",
  name => "Cumulus",
  release => {
    full => "12.5",
    major => "12",
    minor => "5"
  },
  selinux => {
    enabled => false
  }
}

Reply:
Thanks, the root cause is because facter 3 used to report [`os.name => "CumulusLinux"`](https://github.com/puppetlabs/facter/blob/6d7ffc6efdfbc3b1fc79311cdeb4581ac2098d9c/acceptance/lib/facter/acceptance/base_fact_utils.rb#L122)

So `systemd` would be selected as the default due to https://github.com/puppetlabs/puppet/blob/dc0e3baae305a6d25164771661edb742ede656ee/lib/puppet/provider/service/systemd.rb#L33

However, facter 4 reports `os.name => "Cumulus"`, hence this issue. To ensure compatibility with facter 3 and 4, could you add the following to the `systemd` provider:

```
defaultfor 'os.name' => :cumulus, 'os.release.major' => %w[5]
```

I'm assuming cumulus 3 and 4 are EOL?

Reply:
Thanks Josh,
agreed!

Reply:
@dpkgbofh I unresolved this based on my comment:

>  To ensure compatibility with facter 3 and 4, could you add the following to the systemd provider

```
defaultfor 'os.name' => :cumulus, 'os.release.major' => %w[5]
```

At which point, I don't think the change to the `init` provider is needed?